core/bin/external_node/src/main.rs

```diff
@@ -387,46 +379,77 @@ async fn main() -> anyhow::Result<()> {
     .await
     .context("Performing genesis failed")?;
 
-    let (task_handles, stop_sender, health_check_handle) =
-        init_tasks(config.clone(), connection_pool.clone())
-            .await
-            .context("init_tasks")?;
+    let (stop_sender, stop_receiver) = watch::channel::<bool>(false);
 
-    let reorg_detector = ReorgDetector::new(&main_node_url, connection_pool.clone());
-    let reorg_detector_handle = tokio::spawn(reorg_detector.run());
+    let (task_handles, mut consistency_checker_handle, health_check_handle) = init_tasks(
+        config.clone(),
+        connection_pool.clone(),
+        stop_receiver.clone(),
+    )
+    .await
+    .context("init_tasks")?;
+
+    let reorg_detector = ReorgDetector::new(&main_node_url, connection_pool.clone(), stop_receiver);
+    let mut reorg_detector_handle = tokio::spawn(reorg_detector.run());
 
     let particular_crypto_alerts = None;
     let graceful_shutdown = None::<futures::future::Ready<()>>;
     let tasks_allowed_to_finish = false;
-    tokio::select! {
-        _ = wait_for_tasks(task_handles, particular_crypto_alerts, graceful_shutdown, tasks_allowed_to_finish) => {},
-        _ = sigint_receiver => {
-            tracing::info!("Stop signal received, shutting down");
-        },
-        last_correct_batch = reorg_detector_handle => {
-            if let Ok(last_correct_batch) = last_correct_batch {
-                tracing::info!("Performing rollback to block {}", last_correct_batch);
-                shutdown_components(stop_sender, health_check_handle).await;
-                let reverter = BlockReverter::new(
-                    config.required.state_cache_path,
-                    config.required.merkle_tree_path,
-                    None,
-                    connection_pool,
-                    L1ExecutedBatchesRevert::Allowed,
-                );
-                reverter
-                    .rollback_db(last_correct_batch, BlockReverterFlags::all())
-                    .await;
-                tracing::info!("Rollback successfully completed, the node has to restart to continue working");
-                return Ok(());
-            } else {
-                tracing::error!("Reorg detector actor failed");
+
+    let wait_for_tasks = wait_for_tasks(
+        task_handles,
+        particular_crypto_alerts,
+        graceful_shutdown,
+        tasks_allowed_to_finish,
+    );
+
+    tokio::pin!(wait_for_tasks);
+
+    loop {
+        tokio::select! {
+            _ = &mut wait_for_tasks => {
```

Comment:
What happens if the same `select!`ed future is polled twice during multiple loop iterations? Will it panic? Will it resolve immediately? IIRC, [`futures::select!`](https://docs.rs/futures/latest/futures/macro.select.html) doesn't evaluate completed futures, but I'm not sure whether this is the case for `tokio::select!` (Tokio docs make it seem like it's not the case).

Reply:
For the concrete case you are pointing at here, all will be fine, tokio will be driving this function to completion, also you can do much more complicated stuff with it. https://tokio.rs/tokio/tutorial/select#loops

Reply:
I'll take your word for it. Just pointing out that the doc you're referencing confirms that polling `&mut _` future multiple times in a loop results in a panic:

> This example uses some new syntax. The first branch includes `, if !done`. This is a branch precondition. Before explaining how it works, let's look at what happens if the precondition is omitted. Leaving out `, if !done` and running the example results in the following output:
>> thread 'main' panicked at '`async fn` resumed after completion', src/main.rs:1:55
>> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

Reply:
hmmmm